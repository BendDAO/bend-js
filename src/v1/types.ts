import BigNumber from 'bignumber.js';

export type ReserveRatesData = {
  id: string;
  symbol: string;
  paramsHistory: {
    variableBorrowIndex: string;
    liquidityIndex: string;
    timestamp: number;
  }[];
};

export type ReserveSupplyData = {
  totalScaledVariableDebt: string;
  variableBorrowIndex: string;
  variableBorrowRate: string;
  availableLiquidity: string;
  lastUpdateTimestamp: number;
};

export type ReserveData = {
  id: string;
  underlyingAsset: string;
  name: string;
  symbol: string;
  decimals: number;
  isActive: boolean;
  isFrozen: boolean;
  borrowingEnabled: boolean;
  reserveFactor: string;
  bTokenAddress: string;
  debtTokenAddress: string;

  optimalUtilisationRate: string;
  baseVariableBorrowRate: string;
  variableRateSlope1: string;
  variableRateSlope2: string;

  liquidityIndex: string;
  variableBorrowIndex: string;
  variableBorrowRate: string;
  avg30DaysVariableBorrowRate?: string;
  availableLiquidity: string;
  liquidityRate: string;
  avg30DaysLiquidityRate?: string;
  totalScaledVariableDebt: string;
  lastUpdateTimestamp: number;
  price: {
    priceInEth: string;
  };
};

export type ComputedReserveData = {
  utilizationRate: string;
  totalDebt: string;
  totalLiquidity: string;
  supplyAPY: string;
  supplyAPR: string;
  variableBorrowAPY: string;
  variableBorrowAPR: string;
} & ReserveData;

export type Supplies = {
  totalVariableDebt: BigNumber;
  totalLiquidity: BigNumber;
};

export type UserReserveData = {
  scaledBTokenBalance: string;
  scaledVariableDebt: string;
  variableBorrowIndex: string;
  reserve: {
    id: string;
    underlyingAsset: string;
    name: string;
    symbol: string;
    decimals: number;
    lastUpdateTimestamp: number;
  };
};

export type ComputedUserReserve = UserReserveData & {
  underlyingBalance: string;
  underlyingBalanceETH: string;
  underlyingBalanceUSD: string;

  totalBorrows: string;
  totalBorrowsETH: string;
  totalBorrowsUSD: string;
};

export type NftData = {
  id: string;
  underlyingAsset: string;
  name: string;
  symbol: string;
  isActive: boolean;
  isFrozen: boolean;
  baseLTVasCollateral: string;
  liquidationThreshold: string;
  liquidationBonus: string;
  redeemDuration: string;
  auctionDuration: string;
  redeemFine: string;
  bnftAddress: string;

  totalCollateral: string;
  price: {
    priceInEth: string;
  };
};

export type UserNftData = {
  totalCollateral: string;
  nftAsset: {
    id: string;
    underlyingAsset: string;
    name: string;
    symbol: string;
    lastUpdateTimestamp: number;
  };
};

export type ComputedUserNft = UserNftData & {
  underlyingCollateralETH: string;
  underlyingCollateralUSD: string;
};

export type LoanData = {
  id: string;
  loanId: string;
  borrower: string;
  nftAsset: {
    id: string;
    underlyingAsset: string;
    name: string;
    symbol: string;
    lastUpdateTimestamp: number;
  };
  nftTokenId: string;
  reserveAsset: {
    id: string;
    underlyingAsset: string;
    name: string;
    symbol: string;
    decimals: number;
    lastUpdateTimestamp: number;
  };
  scaledAmount: string;
  currentAmount: string;
  state: string;
  bidStartTimestamp: string;
  bidPrice: string;
  bidBorrowAmount: string;
};

export type ComputedLoanData = LoanData & {
  currentAmount: string;
  currentAmountETH: string;
  currentAmountUSD: string;
  healthFactor: string;
};

export type UserSummaryData = {
  id: string;

  totalLiquidityETH: string;
  totalLiquidityUSD: string;

  totalCollateralETH: string;
  totalCollateralUSD: string;

  totalBorrowsETH: string;
  totalBorrowsUSD: string;

  // BEND Token Incentive
  totalRewards: string;
  totalRewardsETH: string;
  totalRewardsUSD: string;

  reservesData: ComputedUserReserve[];
  nftsData: ComputedUserNft[];
  loansData: ComputedLoanData[];
};
