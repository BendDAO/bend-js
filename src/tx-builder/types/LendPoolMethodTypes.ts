import { tEthereumAddress, tStringCurrencyUnits } from '.';

export type LPDepositParamsType = {
  user: tEthereumAddress;
  reserve: tEthereumAddress;
  amount: tStringCurrencyUnits;
  onBehalfOf?: tEthereumAddress;
  referralCode?: string;
};

export type LPWithdrawParamsType = {
  user: tEthereumAddress;
  reserve: tEthereumAddress;
  amount: tStringCurrencyUnits;
  onBehalfOf?: tEthereumAddress;
};

export type LPBorrowParamsType = {
  user: tEthereumAddress;
  reserve: tEthereumAddress;
  amount: tStringCurrencyUnits;
  nftAsset: tEthereumAddress;
  nftTokenId: string;
  onBehalfOf?: tEthereumAddress;
  referralCode?: string;
};

export type LPRepayParamsType = {
  user: tEthereumAddress;
  nftAsset: tEthereumAddress;
  nftTokenId: string;
  reserve: tEthereumAddress;
  amount: tStringCurrencyUnits;
  onBehalfOf?: tEthereumAddress;
};

export type LPAuctionParamsType = {
  user: tEthereumAddress;
  nftAsset: tEthereumAddress;
  nftTokenId: string;
  reserve: tEthereumAddress;
  bidPrice: tStringCurrencyUnits;
  onBehalfOf?: tEthereumAddress;
};

export type LPRedeemParamsType = {
  user: tEthereumAddress;
  nftAsset: tEthereumAddress;
  nftTokenId: string;
  reserve: tEthereumAddress;
  amount: tStringCurrencyUnits;
};

export type LPLiquidateParamsType = {
  user: tEthereumAddress;
  nftAsset: tEthereumAddress;
  nftTokenId: string;
};
