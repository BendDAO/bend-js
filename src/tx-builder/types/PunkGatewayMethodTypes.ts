import { tEthereumAddress, tStringCurrencyUnits } from '.';

export type PunkERC20BorrowParamsType = {
  user: tEthereumAddress;
  reserve: tEthereumAddress;
  amount: tStringCurrencyUnits;
  wpunkAddress: tEthereumAddress;
  punkIndex: string;
  onBehalfOf?: tEthereumAddress;
  referralCode?: string;
};

export type PunkERC20RepayParamsType = {
  user: tEthereumAddress;
  punkIndex: string;
  reserve: tEthereumAddress;
  amount: tStringCurrencyUnits;
};

export type PunkERC20AuctionParamsType = {
  user: tEthereumAddress;
  punkIndex: string;
  reserve: tEthereumAddress;
  bidPrice: tStringCurrencyUnits;
  onBehalfOf?: tEthereumAddress;
};

export type PunkERC20RedeemParamsType = {
  user: tEthereumAddress;
  punkIndex: string;
  reserve: tEthereumAddress;
  amount: tStringCurrencyUnits;
};

export type PunkERC20LiquidateParamsType = {
  user: tEthereumAddress;
  punkIndex: string;
};

export type PunkETHBorrowParamsType = {
  user: tEthereumAddress;
  amount: tStringCurrencyUnits;
  wpunkAddress: tEthereumAddress;
  punkIndex: string;
  onBehalfOf?: tEthereumAddress;
  referralCode?: string;
};

export type PunkETHRepayParamsType = {
  user: tEthereumAddress;
  punkIndex: string;
  amount: tStringCurrencyUnits;
};

export type PunkETHAuctionParamsType = {
  user: tEthereumAddress;
  punkIndex: string;
  bidPrice: tStringCurrencyUnits;
  onBehalfOf?: tEthereumAddress;
};

export type PunkETHRedeemParamsType = {
  user: tEthereumAddress;
  punkIndex: string;
  amount: tStringCurrencyUnits;
};

export type PunkETHLiquidateParamsType = {
  user: tEthereumAddress;
  punkIndex: string;
};
