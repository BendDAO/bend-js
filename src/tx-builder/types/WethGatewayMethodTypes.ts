import { tEthereumAddress, tStringCurrencyUnits } from '.';

export type WETHDepositParamsType = {
  user: tEthereumAddress;
  amount: tStringCurrencyUnits;
  onBehalfOf?: tEthereumAddress;
  referralCode?: string;
};

export type WETHWithdrawParamsType = {
  user: tEthereumAddress;
  amount: tStringCurrencyUnits;
  bTokenAddress: tEthereumAddress;
  onBehalfOf?: tEthereumAddress;
};

export type WETHBorrowParamsType = {
  user: tEthereumAddress;
  amount: tStringCurrencyUnits;
  bTokenAddress: tEthereumAddress;
  nftAsset: tEthereumAddress;
  nftTokenId: string;
  onBehalfOf?: tEthereumAddress;
  referralCode?: string;
};

export type WETHRepayParamsType = {
  user: tEthereumAddress;
  nftAsset: tEthereumAddress;
  nftTokenId: string;
  amount: tStringCurrencyUnits;
};

export type WETHAuctionParamsType = {
  user: tEthereumAddress;
  nftAsset: tEthereumAddress;
  nftTokenId: string;
  bidPrice: tStringCurrencyUnits;
  onBehalfOf?: tEthereumAddress;
};

export type WETHRedeemParamsType = {
  user: tEthereumAddress;
  nftAsset: tEthereumAddress;
  nftTokenId: string;
  amount: tStringCurrencyUnits;
};

export type WETHLiquidateParamsType = {
  user: tEthereumAddress;
  nftAsset: tEthereumAddress;
  nftTokenId: string;
};
