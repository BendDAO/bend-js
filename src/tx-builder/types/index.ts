import { providers, BigNumber, BytesLike, PopulatedTransaction } from 'ethers';

export type tEthereumAddress = string;
export type tStringCurrencyUnits = string; // ex 2.5 eth
export type tStringDecimalUnits = string; // ex 2500000000000000000
export type ENS = string; // something.eth

export enum Market {
  Proto = 'proto',
}

export enum Network {
  mainnet = 'mainnet',
  ropsten = 'ropsten',
  rinkeby = 'rinkeby',
  kovan = 'kovan',
  fork = 'fork',
}

export enum ChainId {
  mainnet = 1,
  ropsten = 3,
  rinkeby = 4,
  kovan = 42,
  fork = 1337,
}
export type ConstantAddressesByNetwork = {
  [network: string]: {
    CRYPTO_PUNKS?: tEthereumAddress;
  };
};

export type IncentivesConfig = {
  INCENTIVES_CONTROLLER: tEthereumAddress;
  INCENTIVES_CONTROLLER_REWARD_TOKEN: tEthereumAddress;
};

export type LendPoolMarketConfig = {
  LEND_POOL: tEthereumAddress;
  WETH_GATEWAY?: tEthereumAddress;
  PUNK_GATEWAY?: tEthereumAddress;
  CRYPTO_PUNKS?: tEthereumAddress;
  FAUCET?: tEthereumAddress;
};

export type LendPoolConfig = {
  [network: string]: {
    [market: string]: LendPoolMarketConfig;
  };
};

export type TxBuilderConfig = {
  lendPool?: LendPoolConfig;
  incentives?: {
    [network: string]: IncentivesConfig;
  };
};

export enum eEthereumTxType {
  ERC20_APPROVAL = 'ERC20_APPROVAL',
  FAUCET_MINT = 'FAUCET_MINT',
  REWARD_ACTION = 'REWARD_ACTION',
  DLP_ACTION = 'DLP_ACTION',
}

export enum ProtocolAction {
  default = 'default',
  deposit = 'deposit',
  withdraw = 'withdraw',
  borrow = 'borrow',
  repay = 'repay',
  auction = 'auction',
  redeem = 'redeem',
  liquidate = 'liquidate',
  depositETH = 'depositETH',
  withdrawETH = 'withdrawETH',
  borrowETH = 'borrwoETH',
  repayETH = 'repayETH',
  auctionETH = 'auctionETH',
  redeemETH = 'redeemETH',
  liquidateETH = 'liquidateETH',
}

export type GasRecommendationType = {
  [action: string]: {
    limit: string;
    recommended: string;
  };
};

export type GeneratedTx = {
  tx: transactionType;
  gas: {
    price: string;
    limit: string;
  };
};

export type transactionType = {
  value?: string;
  from?: string;
  to?: string;
  nonce?: number;
  gasLimit?: BigNumber;
  gasPrice?: BigNumber;
  data?: string;
  chainId?: number;
};

export type AddressModel = {
  ADDRESS_PROVIDER_ADDRESS: tEthereumAddress;
  LENDPOOL_ADDRESS: tEthereumAddress;
  WETH_GATEWAY: tEthereumAddress;
  PUNK_GATEWAY: tEthereumAddress;
  FAUCET: tEthereumAddress;
  INCENTIVES_CONTROLLER: tEthereumAddress;
  INCENTIVES_CONTROLLER_REWARD_TOKEN: tEthereumAddress;
};

export type tCommonContractAddressBetweenMarkets = Pick<
  AddressModel,
  | 'WETH_GATEWAY'
  | 'PUNK_GATEWAY'
  | 'FAUCET'
  | 'INCENTIVES_CONTROLLER'
  | 'INCENTIVES_CONTROLLER_REWARD_TOKEN'
>;

export type tDistinctContractAddressBetweenMarkets = Pick<
  AddressModel,
  'LENDPOOL_ADDRESS'
>;

export type ContractAddresses = {
  [contractName: string]: tEthereumAddress;
};

export type Configuration = {
  network: Network;
  provider: providers.Provider;
};

export type EthereumTransactionTypeExtended = {
  txType: eEthereumTxType;
  tx: () => Promise<transactionType>;
  gas: GasResponse;
};

export type TransactionGenerationMethod = {
  rawTxMethod: () => Promise<PopulatedTransaction>;
  from: tEthereumAddress;
  value?: string;
  gasSurplus?: number;
  action?: ProtocolAction;
};

export type TransactionGasGenerationMethod = {
  txCallback: () => Promise<transactionType>;
  action?: ProtocolAction;
};

export type GasType = {
  gasLimit: string | undefined;
  gasPrice: string;
};
export type GasResponse = (force?: boolean) => Promise<GasType | null>;

export type TokenMetadataType = {
  name: string;
  symbol: string;
  decimals: number;
  address: string;
};

export type DefaultProviderKeys = {
  etherscan?: string;
  infura?: string;
  alchemy?: string;
};

export type CommonConfigType = {
  [network: string]: tCommonContractAddressBetweenMarkets;
};

export type LendPoolConfigType = {
  [pool: string]: {
    [network: string]: tDistinctContractAddressBetweenMarkets;
  };
};

export type EnabledNetworksType = {
  lendPool: {
    [market: string]: Network[];
  };
  wethGateway: Network[];
  punkGateway: Network[];
  faucet: Network[];
  incentivesController: Network[];
};

export type PermitSignature = {
  amount: tStringCurrencyUnits;
  deadline: string;
  v: number;
  r: BytesLike;
  s: BytesLike;
};
