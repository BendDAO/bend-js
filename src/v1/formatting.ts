import {
  BigNumberValue,
  valueToBigNumber,
  valueToZDBigNumber,
  normalize,
} from '../helpers/bignumber';
import { calculateAverageRate, LTV_PRECISION } from '../helpers/pool-math';
import { RAY, rayPow } from '../helpers/ray-math';
import {
  ComputedUserReserve,
  ReserveData,
  UserReserveData,
  UserSummaryData,
  ReserveRatesData,
  ComputedReserveData,
  LoanData,
  ComputedLoanData,
  NftData,
} from './types';
import {
  ETH_DECIMALS,
  RAY_DECIMALS,
  SECONDS_PER_YEAR,
  USD_DECIMALS,
  BEND_DECIMALS,
} from '../helpers/constants';
import {
  calculateReserveDebt,
  ComputedUserNft,
  computeRawUserSummaryData,
  UserIncentive,
  UserNftData,
} from '..';
import './computations';

export function formatUserSummaryData(
  poolReservesData: ReserveData[],
  rawUserReserves: UserReserveData[],
  poolNftsData: NftData[],
  rawUserNfts: UserNftData[],
  rawLoanDatas: LoanData[],
  rawUserIncentives: UserIncentive[],
  userId: string,
  usdPriceEth: BigNumberValue,
  currentTimestamp: number
): UserSummaryData {
  const userData = computeRawUserSummaryData(
    poolReservesData,
    rawUserReserves,
    poolNftsData,
    rawUserNfts,
    rawLoanDatas,
    rawUserIncentives,
    userId,
    usdPriceEth,
    currentTimestamp
  );

  const userReservesData = userData.reservesData.map(
    ({ reserve, ...userReserve }): ComputedUserReserve => {
      return {
        ...userReserve,
        reserve: {
          ...reserve,
        },

        scaledBTokenBalance: normalize(
          userReserve.scaledBTokenBalance,
          reserve.decimals
        ),

        scaledVariableDebt: normalize(
          userReserve.scaledVariableDebt,
          reserve.decimals
        ),
        variableBorrowIndex: normalize(
          userReserve.variableBorrowIndex,
          RAY_DECIMALS
        ),

        underlyingBalance: normalize(
          userReserve.underlyingBalance,
          reserve.decimals
        ),
        underlyingBalanceETH: normalize(
          userReserve.underlyingBalanceETH,
          ETH_DECIMALS
        ),
        underlyingBalanceUSD: normalize(
          userReserve.underlyingBalanceUSD,
          USD_DECIMALS
        ),

        totalBorrows: normalize(userReserve.totalBorrows, reserve.decimals),
        totalBorrowsETH: normalize(userReserve.totalBorrowsETH, ETH_DECIMALS),
        totalBorrowsUSD: normalize(userReserve.totalBorrowsUSD, USD_DECIMALS),
      };
    }
  );

  const userNftsData = userData.nftsData.map(
    ({ nftAsset, ...userNft }): ComputedUserNft => {
      return {
        ...userNft,
        nftAsset: {
          ...nftAsset,
        },

        underlyingCollateralETH: normalize(
          userNft.underlyingCollateralETH,
          ETH_DECIMALS
        ),
        underlyingCollateralUSD: normalize(
          userNft.underlyingCollateralUSD,
          USD_DECIMALS
        ),
      };
    }
  );

  const loansData = userData.loansData.map(
    ({ nftAsset, ...loanData }): ComputedLoanData => {
      return {
        ...loanData,
        nftAsset: {
          ...nftAsset,
        },

        currentAmount: normalize(
          loanData.currentAmount,
          loanData.reserveAsset.decimals
        ),
        currentAmountETH: normalize(loanData.currentAmountETH, ETH_DECIMALS),
        currentAmountUSD: normalize(loanData.currentAmountUSD, USD_DECIMALS),
      };
    }
  );

  const incentivesData = userData.incentivesData.map(
    ({ asset, ...incentiveData }): UserIncentive => {
      return {
        ...incentiveData,
        asset: {
          ...asset,
        },

        lifetimeRewards: normalize(
          incentiveData.lifetimeRewards,
          BEND_DECIMALS
        ),
        reward: normalize(incentiveData.reward, BEND_DECIMALS),
      };
    }
  );

  return {
    id: userData.id,

    totalLiquidityETH: normalize(userData.totalLiquidityETH, ETH_DECIMALS),
    totalLiquidityUSD: normalize(userData.totalLiquidityUSD, USD_DECIMALS),

    totalCollateralETH: normalize(userData.totalCollateralETH, ETH_DECIMALS),
    totalCollateralUSD: normalize(userData.totalCollateralUSD, USD_DECIMALS),

    totalBorrowsETH: normalize(userData.totalBorrowsETH, ETH_DECIMALS),
    totalBorrowsUSD: normalize(userData.totalBorrowsUSD, USD_DECIMALS),

    totalRewards: normalize(userData.totalRewards, BEND_DECIMALS),

    reservesData: userReservesData,
    nftsData: userNftsData,
    loansData: loansData,
    incentivesData: incentivesData,
  };
}

export function formatReserves(
  reserves: ReserveData[],
  currentTimestamp?: number,
  reserveIndexes30DaysAgo?: ReserveRatesData[]
): ComputedReserveData[] {
  return reserves.map((reserve) => {
    const reserve30DaysAgo = reserveIndexes30DaysAgo?.find(
      (res) => res.id === reserve.id
    )?.paramsHistory?.[0];

    const availableLiquidity = normalize(
      reserve.availableLiquidity,
      reserve.decimals
    );

    const { totalVariableDebt } = calculateReserveDebt(
      reserve,
      currentTimestamp || reserve.lastUpdateTimestamp
    );

    const totalDebt = valueToBigNumber(totalVariableDebt);

    const totalLiquidity = totalDebt.plus(availableLiquidity).toString();
    const utilizationRate =
      totalLiquidity !== '0'
        ? totalDebt.dividedBy(totalLiquidity).toString()
        : '0';

    const supplyAPY = rayPow(
      valueToZDBigNumber(reserve.liquidityRate)
        .dividedBy(SECONDS_PER_YEAR)
        .plus(RAY),
      SECONDS_PER_YEAR
    ).minus(RAY);

    const variableBorrowAPY = rayPow(
      valueToZDBigNumber(reserve.variableBorrowRate)
        .dividedBy(SECONDS_PER_YEAR)
        .plus(RAY),
      SECONDS_PER_YEAR
    ).minus(RAY);

    return {
      ...reserve,
      totalVariableDebt,
      totalLiquidity,
      availableLiquidity,
      utilizationRate,
      totalDebt: totalDebt.toString(),
      price: {
        ...reserve.price,
        priceInEth: normalize(reserve.price.priceInEth, ETH_DECIMALS),
      },
      reserveFactor: normalize(reserve.reserveFactor, LTV_PRECISION),

      variableBorrowAPR: normalize(reserve.variableBorrowRate, RAY_DECIMALS),
      variableBorrowAPY: normalize(variableBorrowAPY, RAY_DECIMALS),

      avg30DaysVariableBorrowRate: reserve30DaysAgo
        ? calculateAverageRate(
            reserve30DaysAgo.variableBorrowIndex,
            reserve.variableBorrowIndex,
            reserve30DaysAgo.timestamp,
            reserve.lastUpdateTimestamp
          )
        : undefined,

      avg30DaysLiquidityRate: reserve30DaysAgo
        ? calculateAverageRate(
            reserve30DaysAgo.liquidityIndex,
            reserve.liquidityIndex,
            reserve30DaysAgo.timestamp,
            reserve.lastUpdateTimestamp
          )
        : undefined,

      supplyAPR: normalize(reserve.liquidityRate, RAY_DECIMALS),
      supplyAPY: normalize(supplyAPY, RAY_DECIMALS),

      liquidityIndex: normalize(reserve.liquidityIndex, RAY_DECIMALS),
      liquidityRate: normalize(reserve.liquidityRate, RAY_DECIMALS),

      totalScaledVariableDebt: normalize(
        reserve.totalScaledVariableDebt,
        reserve.decimals
      ),
      variableBorrowIndex: normalize(reserve.variableBorrowIndex, RAY_DECIMALS),
      variableBorrowRate: normalize(reserve.variableBorrowRate, RAY_DECIMALS),
    };
  });
}
