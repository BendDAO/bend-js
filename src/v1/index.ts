export * from './types';
export * from './computations';
export * from './formatting';
