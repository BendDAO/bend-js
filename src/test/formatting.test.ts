import { ReserveData } from '../v1/types';
import { formatReserves, formatUserSummaryData } from '../v1/formatting';
import BigNumber from 'bignumber.js';
import {
  mockReserve,
  mockUserReserve,
  mockNft,
  mockUserNft,
  mockUserIncentive,
  mockLoan,
  mockUserId,
} from './mock-data';

describe('computations and formattings', () => {
  describe('formatUserSummaryData', () => {
    const formattedMockReserve = formatUserSummaryData(
      [mockReserve],
      [mockUserReserve],
      [mockNft],
      [mockUserNft],
      [mockLoan],
      [mockUserIncentive],
      mockUserId,
      '598881655557838',
      mockUserReserve.reserve.lastUpdateTimestamp + 2000
    );
    expect(formattedMockReserve).toMatchSnapshot();
  });

  describe('formatReserves', () => {
    it('should return plausible results', () => {
      const formattedMockReserve = formatReserves(
        [mockReserve],
        mockReserve.lastUpdateTimestamp + 2000
      )[0];
      expect(formattedMockReserve).toMatchSnapshot();
    });

    it('should allow omitting timestamp', () => {
      const formattedMockReserve = formatReserves([mockReserve])[0];
      expect(formattedMockReserve).toMatchSnapshot();
    });

    /**
     * Whenever we add a new asset there#s a chance that an asset has no paramsHistory from 30days ago
     * We should not throw if that's the case, but just ignore it
     */
    it("should not error When 30dago reserves doesn't contain paramsHistory", () => {
      formatReserves([mockReserve], mockReserve.lastUpdateTimestamp + 2000, [
        mockReserve as any,
      ]);
      formatReserves([mockReserve], mockReserve.lastUpdateTimestamp + 2000, [
        { ...mockReserve, paramsHistory: [] } as any,
      ]);
      // would be wrong if any of the above threw
      expect(true).toBe(true);
    });

    it('should return proper values for new reserves', () => {
      const newReserve: Partial<ReserveData> = {
        availableLiquidity: '0',
        totalScaledVariableDebt: '0',
      };
      const formattedMockReserve = formatReserves(
        [
          {
            ...mockReserve,
            ...newReserve,
          },
        ],
        mockReserve.lastUpdateTimestamp
      )[0];
      expect(formattedMockReserve.utilizationRate).toBe('0');
    });

    it('should increase over time', () => {
      /**
       * tests against a regression which switched two dates
       */
      const first = formatReserves(
        [mockReserve],
        mockReserve.lastUpdateTimestamp + 1,
        [mockReserve as any]
      )[0];
      const second = formatReserves(
        [mockReserve],
        mockReserve.lastUpdateTimestamp + 2,
        [mockReserve as any]
      )[0];

      expect(new BigNumber(second.totalDebt).gte(first.totalDebt)).toBe(true);
    });
  });
});
