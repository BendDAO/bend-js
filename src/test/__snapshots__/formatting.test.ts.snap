// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
Object {
  "id": "0x20000b9b01e93a39db9d286e9264eff7f2af16e9",
  "incentivesData": Array [
    Object {
      "asset": Object {
        "emissionPerSecond": 0,
        "id": "0xd800e97ae32b06c1e89ca5126c7bf6aef89d6b240xd800e97ae32b06c1e89ca5126c7bf6aef89d6b24",
        "index": 0,
      },
      "lastUpdateTimestamp": 1641228846,
      "lifetimeRewards": "1545.269812517967062645",
      "reward": "1545.269812517967062645",
    },
  ],
  "loansData": Array [
    Object {
      "bidBorrowAmount": "0",
      "bidPrice": "0",
      "bidStartTimestamp": "0",
      "bidderAddress": "0x00",
      "borrower": "0x20000b9b01e93a39db9d286e9264eff7f2af16e9",
      "currentAmount": "2.096644626429035316",
      "currentAmountETH": "2.096644626429035316",
      "currentAmountUSD": "3500.9331258879",
      "healthFactor": "0.02129116181030162794",
      "id": "50xe55870ebb007a50b0dfabadb1a21e4bfcee5299b",
      "lastUpdateTimestamp": 1641216645,
      "loanId": "5",
      "nftAsset": Object {
        "id": "0x74e4418a41169fb951ca886976ccd8b36968c4ab0xe55870ebb007a50b0dfabadb1a21e4bfcee5299b",
        "lastUpdateTimestamp": 1642218270,
        "name": "Wrapped Cryptopunks",
        "symbol": "WPUNKS",
        "underlyingAsset": "0x74e4418a41169fb951ca886976ccd8b36968c4ab",
      },
      "nftTokenId": "1337",
      "reserveAsset": Object {
        "decimals": 18,
        "id": "0xad1908f909b5c5d2b1032a215d611773f26f089f0xe55870ebb007a50b0dfabadb1a21e4bfcee5299b",
        "lastUpdateTimestamp": 1642311951,
        "name": "Wrapped Ether",
        "symbol": "WETH",
        "underlyingAsset": "0xad1908f909b5c5d2b1032a215d611773f26f089f",
      },
      "scaledAmount": "2012908826592418942",
      "state": "Active",
    },
  ],
  "nftsData": Array [
    Object {
      "nftAsset": Object {
        "id": "0x74e4418a41169fb951ca886976ccd8b36968c4ab0xe55870ebb007a50b0dfabadb1a21e4bfcee5299b",
        "lastUpdateTimestamp": 1642218270,
        "name": "Wrapped Cryptopunks",
        "symbol": "WPUNKS",
        "underlyingAsset": "0x74e4418a41169fb951ca886976ccd8b36968c4ab",
      },
      "totalCollateral": "3",
      "underlyingCollateralETH": "0.1488",
      "underlyingCollateralUSD": "248.4631122344",
    },
  ],
  "reservesData": Array [
    Object {
      "reserve": Object {
        "decimals": 18,
        "id": "0xad1908f909b5c5d2b1032a215d611773f26f089f0xe55870ebb007a50b0dfabadb1a21e4bfcee5299b",
        "lastUpdateTimestamp": 1642159201,
        "name": "Wrapped Ether",
        "symbol": "WETH",
        "underlyingAsset": "0xad1908f909b5c5d2b1032a215d611773f26f089f",
      },
      "scaledBTokenBalance": "1.119602730417571034",
      "scaledVariableDebt": "2.012908826592418942",
      "totalBorrows": "2.096644626429035316",
      "totalBorrowsETH": "2.096644626429035316",
      "totalBorrowsUSD": "3500.9331258879",
      "underlyingBalance": "1.159297638269989947",
      "underlyingBalanceETH": "1.159297638269989947",
      "underlyingBalanceUSD": "1935.770828028",
      "variableBorrowIndex": "1.01007509787777865131",
    },
  ],
  "totalBorrowsETH": "4.193289252858070632",
  "totalBorrowsUSD": "7001.8662517758",
  "totalCollateralETH": "0.1488",
  "totalCollateralUSD": "248.4631122344",
  "totalLiquidityETH": "1.159297638269989947",
  "totalLiquidityUSD": "1935.770828028",
  "totalRewards": "1545.269812517967062645",
}
`;

exports[`computations and formattings formatReserves should allow omitting timestamp 1`] = `
Object {
  "availableLiquidity": "30.176110452260548839",
  "avg30DaysLiquidityRate": undefined,
  "avg30DaysVariableBorrowRate": undefined,
  "bTokenAddress": "0x1bbce5469b8bcc5078ae2398476350936d1393af",
  "baseVariableBorrowRate": "0",
  "borrowingEnabled": true,
  "debtTokenAddress": "0xe42f3a56f89546a2596b88cff08234c5eea304b7",
  "decimals": 18,
  "id": "0xad1908f909b5c5d2b1032a215d611773f26f089f0xe55870ebb007a50b0dfabadb1a21e4bfcee5299b",
  "isActive": true,
  "isFrozen": false,
  "lastUpdateTimestamp": 1642159201,
  "liquidityIndex": "1.03541834282012424654",
  "liquidityRate": "0.54998284644218217077",
  "name": "Wrapped Ether",
  "optimalUtilisationRate": "650000000000000000000000000",
  "price": Object {
    "priceInEth": "1",
  },
  "reserveFactor": "0.1",
  "supplyAPR": "0.54998284644218217077",
  "supplyAPY": "0.73322327835433494494",
  "symbol": "WETH",
  "totalDebt": "198.982640816212378612",
  "totalLiquidity": "229.158751268472927451",
  "totalScaledVariableDebt": "191.044198058136805282",
  "totalVariableDebt": "198.982640816212378612",
  "underlyingAsset": "0xad1908f909b5c5d2b1032a215d611773f26f089f",
  "utilizationRate": "0.86831787882755800028",
  "variableBorrowAPR": "0.70376536807873714367",
  "variableBorrowAPY": "1.02134950518465855846",
  "variableBorrowIndex": "1.04155291204216429422",
  "variableBorrowRate": "0.70376536807873714367",
  "variableRateSlope1": "80000000000000000000000000",
  "variableRateSlope2": "1000000000000000000000000000",
}
`;

exports[`computations and formattings formatReserves should return plausible results 1`] = `
Object {
  "availableLiquidity": "30.176110452260548839",
  "avg30DaysLiquidityRate": undefined,
  "avg30DaysVariableBorrowRate": undefined,
  "bTokenAddress": "0x1bbce5469b8bcc5078ae2398476350936d1393af",
  "baseVariableBorrowRate": "0",
  "borrowingEnabled": true,
  "debtTokenAddress": "0xe42f3a56f89546a2596b88cff08234c5eea304b7",
  "decimals": 18,
  "id": "0xad1908f909b5c5d2b1032a215d611773f26f089f0xe55870ebb007a50b0dfabadb1a21e4bfcee5299b",
  "isActive": true,
  "isFrozen": false,
  "lastUpdateTimestamp": 1642159201,
  "liquidityIndex": "1.03541834282012424654",
  "liquidityRate": "0.54998284644218217077",
  "name": "Wrapped Ether",
  "optimalUtilisationRate": "650000000000000000000000000",
  "price": Object {
    "priceInEth": "1",
  },
  "reserveFactor": "0.1",
  "supplyAPR": "0.54998284644218217077",
  "supplyAPY": "0.73322327835433494494",
  "symbol": "WETH",
  "totalDebt": "198.991522108389090691",
  "totalLiquidity": "229.16763256064963953",
  "totalScaledVariableDebt": "191.044198058136805282",
  "totalVariableDebt": "198.991522108389090691",
  "underlyingAsset": "0xad1908f909b5c5d2b1032a215d611773f26f089f",
  "utilizationRate": "0.86832298211103444039",
  "variableBorrowAPR": "0.70376536807873714367",
  "variableBorrowAPY": "1.02134950518465855846",
  "variableBorrowIndex": "1.04155291204216429422",
  "variableBorrowRate": "0.70376536807873714367",
  "variableRateSlope1": "80000000000000000000000000",
  "variableRateSlope2": "1000000000000000000000000000",
}
`;
